### ISSUE TITLE
Client unable to login while upgrading to any version above 17.04

### ISSUE THREAD

**Issue by kalabantas** [processone/ejabberd#2442]
> What version of ejabberd are you using?

Any version upgrade from 17.04 using the deb files.

> What operating system (version) are you using?

Ubuntu Trusty

> How did you install ejabberd (source, package, distribution)?

Deb file

> What did not work as expected?

Whenever I try to login I get the error 
2018-05-30 02:04:09.641 [error] <0.641.0>@base64:decode_binary:212 CRASH REPORT Process <0.641.0> with 0 neighbours exited with reason: no match of right hand value false in base64:decode_binary/2 line 212 in p1_server:terminate/7 line 878
2018-05-30 02:04:09.641 [error] <0.637.0>@base64:decode_binary:212 Supervisor ejabberd_c2s_sup had child undefined started with {ejabberd_c2s,start_link,undefined} at <0.641.0> exit with reason no match of right hand value false in base64:decode_binary/2 line 212 in context child_terminated

> was the unexpected behavior? What was the expected result?

Client login.

The client logs in normally on version 17.04 and below. Using the same config on any version above 17.04 gives this error from different clients.
I upgraded erl to v 20, upgraded openssl,  tried different ssl and starttls configurations, no change.

**Comment by zinid** [comment]
There should be more info in `crash.log`

**Comment by kalabantas** [comment]
Here is the crash.log 
** Generic server <0.1243.0> terminating 
** Last message in was {'$gen_event',{xmlstreamelement,{xmlel,<<"auth">>,[{<<"xmlns">>,<<"urn:ietf:params:xml:ns:xmpp-sasl">>},{<<"xmlns:ga">>,<<"http://www.google.com/talk/protocol/auth">>},{<<"mechanism">>,<<"PLAIN">>},{<<"ga:client-uses-full-bind-result">>,<<"true">>}],[{xmlcdata,<<"AHhuLS1tZ2JrajlmADUxMTUxMjU1">>}]}}}
** When Server state == #{shaper => c2s_shaper,tls_verify => false,mgmt_stanzas_in => 0,access => c2s,lang => <<"en">>,resource => <<>>,socket => {socket_state,fast_tls,{tlssock,#Port<0.23361>,#Ref<0.2995613210.1760165889.28488>},infinity,#Ref<0.2995613210.1760165889.28479>,{maxrate,1000,990.0374244513326,1527665555408235},undefined},user => <<>>,stream_state => wait_for_sasl_request,stream_timeout => {30000,-576430232854},pres_a => {0,nil},lserver => <<"domain.LTD">>,socket_monitor => #Ref<0.2995613210.1760034817.28477>,stream_version => {1,0},mgmt_state => inactive,stream_direction => in,mod => ejabberd_c2s,mgmt_max_timeout => 300,xmlns => <<"jabber:client">>,stream_encrypted => true,tls_required => true,stream_authenticated => false,codec_options => [ignore_els],stream_id => <<"12750056827606273017">>,mgmt_resend => false,tls_enabled => false,mgmt_queue_type => ram,stream_header_sent => true,mgmt_max_queue => 5000,csi_state => active,owner => <0.1243.0>,zlib => false,mgmt_stanzas_req => 0,mgmt_ack_timeout => 60000,csi_queue => {0,#{}},stream_restarted => true,tls_options => [compression_none,{protocol_options,<<"no_sslv3">>},{certfile,<<"/opt/ejabberd/crt/mycert.pem">>}],server => <<"domain.LTD">>,ip => {{ip1,ip2,ip3,ip4},58928},mgmt_stanzas_out => 0,stream_compressed => false,mgmt_timeout => 300}
** Reason for termination == 
** {{badmatch,false},[{base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},{ejabberd_auth,is_password_scram_valid,2,[{file,"src/ejabberd_auth.erl"},{line,664}]},{ejabberd_auth,'-check_password_with_authmodule/6-fun-0-',8,[{file,"src/ejabberd_auth.erl"},{line,222}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{cyrsasl_plain,mech_step,2,[{file,"src/cyrsasl_plain.erl"},{line,55}]},{cyrsasl,server_step,2,[{file,"src/cyrsasl.erl"},{line,209}]},{xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,809}]},{xmpp_stream_in,handle_info,2,[{file,"src/xmpp_stream_in.erl"},{line,355}]}]}
2018-05-30 10:32:35 =CRASH REPORT====
  crasher:
    initial call: xmpp_stream_in:init/1
    pid: <0.1243.0>
    registered_name: []
    exception exit: {{{badmatch,false},[{base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},{ejabberd_auth,is_password_scram_valid,2,[{file,"src/ejabberd_auth.erl"},{line,664}]},{ejabberd_auth,'-check_password_with_authmodule/6-fun-0-',8,[{file,"src/ejabberd_auth.erl"},{line,222}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{cyrsasl_plain,mech_step,2,[{file,"src/cyrsasl_plain.erl"},{line,55}]},{cyrsasl,server_step,2,[{file,"src/cyrsasl.erl"},{line,209}]},{xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,809}]},{xmpp_stream_in,handle_info,2,[{file,"src/xmpp_stream_in.erl"},{line,355}]}]},[{p1_server,terminate,7,[{file,"src/p1_server.erl"},{line,878}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]}
    ancestors: [ejabberd_c2s_sup,ejabberd_sup,<0.66.0>]
    message_queue_len: 0
    messages: []
    links: [<0.637.0>]
    dictionary: [{already_terminated,true},{rand_seed,{#{bits => 58,jump => #Fun<rand.8.15449617>,next => #Fun<rand.5.15449617>,type => exrop,uniform => #Fun<rand.6.15449617>,uniform_n => #Fun<rand.7.15449617>,weak_low_bits => 1},[75746081593938948|274824863503989334]}},{'$internal_queue_len',0}]
    trap_exit: false
    status: running
    heap_size: 2586
    stack_size: 27
    reductions: 23118
  neighbours:
2018-05-30 10:32:35 =SUPERVISOR REPORT====
     Supervisor: {local,ejabberd_c2s_sup}
     Context:    child_terminated
     Reason:     {{badmatch,false},[{base64,decode_binary,2,[{file,"base64.erl"},{line,212}]},{ejabberd_auth,is_password_scram_valid,2,[{file,"src/ejabberd_auth.erl"},{line,664}]},{ejabberd_auth,'-check_password_with_authmodule/6-fun-0-',8,[{file,"src/ejabberd_auth.erl"},{line,222}]},{lists,foldl,3,[{file,"lists.erl"},{line,1263}]},{cyrsasl_plain,mech_step,2,[{file,"src/cyrsasl_plain.erl"},{line,55}]},{cyrsasl,server_step,2,[{file,"src/cyrsasl.erl"},{line,209}]},{xmpp_stream_in,process_sasl_request,2,[{file,"src/xmpp_stream_in.erl"},{line,809}]},{xmpp_stream_in,handle_info,2,[{file,"src/xmpp_stream_in.erl"},{line,355}]}]}
     Offender:   [{pid,<0.1243.0>},{id,undefined},{mfargs,{ejabberd_c2s,start_link,undefined}},{restart_type,temporary},{shutdown,1000},{child_type,worker}]

**Comment by zinid** [comment]
What auth backend? Mysql?

**Comment by kalabantas** [comment]
Yes the autth backend is MySql

**Comment by zinid** [comment]
You may want to select `users` table for the row matching the user trying to log in. Maybe you may find something weird there. If in doubt you can hide private data and post here.

**Comment by kalabantas** [comment]
Sorry, that was a wrong deduction from my side. Both users can login to the system. It was a caching issue.
The issue is what zinid expected. there were some characters returning in the query in the serverkey and salt fields. when I removed them, the user could login again. That should not be the expected tough, as I have stated in the ejabberd.yml that the password is plain, so these fields should not be returned in the query. I guess that was the default in 17.04, and since then, the query changed.
Many thans
